Mutual TLS Between Choreo Connect and Backend (#4771)
wso2/docs-apim: en/docs/deploy-and-publish/deploy-on-gateway/choreo-connect/security/tls/mutual-tls-between-gateway-and-backend.md

# Mutual TLS Between Choreo Connect and Backend

In Mutual TLS, in addition to the client verifying the server, the server also verifies the client. By having Mutual TLS between Choreo Connect and the backend, both of the following security measures are applied.

> - The router in Choreo Connect only creates connections with a verified set of backend servers

Reviewer: ```suggestion
- The router in Choreo Connect only creates connections with a verified set of backend servers.
```

> - Backend servers would only accept connections by a previously introduced set of router instanaces

Reviewer: ```suggestion
- Backend servers would only accept connections by a previously introduced set of router instances.
```

## Export the Certificates

!!! Info
    **Before you begin**

>     For TLS, if the backend cert has not been added to the Choreo Connect Router truststore already, use the following commands to create a key and a certificate for the backend, and add the certificate to the Router truststore.

Reviewer: ```suggestion
    For TLS, if the backend certificate has not been added to the Choreo Connect Router truststore already, use the following commands to create a key and a certificate for the backend, and add the certificate to the Router truststore.
```

>     1. Generate a private key for the backend

Reviewer: ```suggestion
    1. Generate a private key for the backend.
```

        ```
        openssl genpkey -out backend.key -algorithm RSA -pkeyopt rsa_keygen_bits:2048
        ```

>     2. Create the certificate

Reviewer: ```suggestion
    2. Create the certificate.
```

>         The following command creates a self-signed certificate for the backend. Note the CN and SAN values. Having both in the backend cert is necessary for the router to complete hostname verification when creating a connection with the backend. 

Reviewer: ```suggestion
        The following command creates a self-signed certificate for the backend. Note the CN and SAN values. Having both in the backend certificate is necessary for the router to complete hostname verification when creating a connection with the backend. 
```

        ```
        openssl req -x509 -new -key backend.key -out backend.pem -subj "/CN=example.com" -reqexts SAN -extensions SAN -config <(cat /etc/ssl/openssl.cnf <(printf "[SAN]\nsubjectAltName=DNS:example.com,DNS:www.example.com")) 
        ```

>     3. Add the cert in `backend.pem` to the router as decribed in [Adding Certificates to Specific Clusters]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/security/tls/transport-security/#adding-certificates-to-specific-clusters).

Reviewer: ```suggestion
    3. Add the certificate in `backend.pem` to the router as described in [Adding Certificates to Specific Clusters]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/security/tls/transport-security/#adding-certificates-to-specific-clusters).
```

>     3. Add the cert in `backend.pem` to the router as decribed in [Add a Certificate to Choreo Connect Router as a Trusted Certificate]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/security/tls/transport-security/#adding-certificates-to-specific-clusters).

Reviewer 2: Add a Certificate to Choreo Connect Router as a Trusted Certificate should be changed to Adding Certificates to Specific Clusters

Author: Sorry, I missed it. Fixed with https://github.com/wso2/docs-apim/pull/4771/commits/8d6baedee81ebe9ac8aab3d4467c0af38cbd4d6c

>     3. Add the cert in `backend.pem` to the router as decribed in [Add a Certificate to Choreo Connect Router as a Trusted Certificate]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/security/tls/transport-security/#add-a-certificate-to-choreo-connect-router-as-a-trusted-certificate).

Reviewer 2: Shall we point to this - https://apim.docs.wso2.com/en/latest/deploy-and-publish/deploy-on-gateway/choreo-connect/security/transport-security/#adding-certificates-to-specific-clusters?

Basically cert of the bkend can be given in 2 ways. With the project (with APIM by uploading the cert) and the way you have mentioned. Shall we recommend cert per cluster?

Author: fixed with https://github.com/wso2/docs-apim/pull/4771/commits/e3ec7e4064b7d6d4cb7da59269da4d9cc38897a8

    Important: The ciphers configured [here]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/security/transport-security/#configuration) must match the ciphers supported by your backend.

> For Mutual TLS, in additon to the [configurations done for TLS]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/security/tls/transport-security/#adding-certificates-to-specific-clusters), the cert of the router must also exist in the backend truststore.

Reviewer: ```suggestion
For Mutual TLS, in addition to the [configurations done for TLS]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/security/tls/transport-security/#adding-certificates-to-specific-clusters), the certificate of the router must also exist in the backend truststore.
```

### Add the Router Certificate to the Backend Truststore

> - If the backend truststore is in `.jks` format use:

Reviewer: ```suggestion
- If the backend truststore is in `.jks` format, use:
```
